crates/milli/src/fields_ids_map/metadata.rs

```diff
@@ -106,76 +121,247 @@ impl Metadata {
+    pub fn is_faceted(&self, rules: &[FilterableAttributesRule]) -> bool {
+        if self.is_distinct() || self.is_sortable() || self.is_asc_desc() {
+            return true;
+        }
+
+        let features = self.filterable_attributes_features(rules);
+        if features.is_filterable() || features.is_facet_searchable() {
+            return true;
+        }
+
+        false
+    }
+
+    pub fn require_facet_level_database(&self, rules: &[FilterableAttributesRule]) -> bool {
+        let features = self.filterable_attributes_features(rules);
+
+        self.is_sortable() || self.is_asc_desc() || features.is_filterable_comparison()
+    }
 }
 
 #[derive(Debug, Clone)]
 pub struct MetadataBuilder {
-    searchable_attributes: Vec<String>,
-    filterable_attributes: HashSet<String>,
+    searchable_attributes: Option<Vec<String>>,
+    filterable_attributes: Vec<FilterableAttributesRule>,
     sortable_attributes: HashSet<String>,
     localized_attributes: Option<Vec<LocalizedAttributesRule>>,
+    distinct_attribute: Option<String>,
+    asc_desc_attributes: HashSet<String>,
 }
 
 impl MetadataBuilder {
     pub fn from_index(index: &Index, rtxn: &RoTxn) -> Result<Self> {
-        let searchable_attributes =
-            index.searchable_fields(rtxn)?.into_iter().map(|s| s.to_string()).collect();
-        let filterable_attributes = index.filterable_fields(rtxn)?;
+        let searchable_attributes = match index.user_defined_searchable_fields(rtxn)? {
+            Some(fields) => Some(fields.into_iter().map(|s| s.to_string()).collect()),
+            None => None,
+        };
+        let filterable_attributes = index.filterable_attributes_rules(rtxn)?;
         let sortable_attributes = index.sortable_fields(rtxn)?;
         let localized_attributes = index.localized_attributes_rules(rtxn)?;
+        let distinct_attribute = index.distinct_field(rtxn)?.map(|s| s.to_string());
+        let asc_desc_attributes = index.asc_desc_fields(rtxn)?;
 
-        Ok(Self {
+        Ok(Self::_new(
             searchable_attributes,
             filterable_attributes,
             sortable_attributes,
             localized_attributes,
-        })
+            distinct_attribute,
+            asc_desc_attributes,
+        ))
     }
 
+    #[cfg(test)]
+    /// Build a new `MetadataBuilder` from the given parameters.
+    ///
+    /// This is used for testing, prefer using `MetadataBuilder::from_index` instead.
     pub fn new(
-        searchable_attributes: Vec<String>,
-        filterable_attributes: HashSet<String>,
+        searchable_attributes: Option<Vec<String>>,
+        filterable_attributes: Vec<FilterableAttributesRule>,
+        sortable_attributes: HashSet<String>,
+        localized_attributes: Option<Vec<LocalizedAttributesRule>>,
+        distinct_attribute: Option<String>,
+        asc_desc_attributes: HashSet<String>,
+    ) -> Self {
+        Self::_new(
```

Comment:
I don't understand the need for `_new` here. Can you inline it in `new` and replace its callers by callers to `new`?

Reply:
I don't want new to be accessible from anything else that test

Reply:
Ah. I see. I don't think it would cause much harm to remove the `cfg[test]` here, but if you want to keep it, maybe you could highlight this with a comment on top of `_new`:

```rust
    // Note: `pub fn new` is #[cfg(test)], this function is private
    fn _new(
```